vault/logical_system.go

```diff
@@ -3415,25 +3472,98 @@ func (b *SystemBackend) pathInternalUIMountsRead(ctx context.Context, req *logic
 	resp.Data["secret"] = secretMounts
 	resp.Data["auth"] = authMounts
 
+	var acl *ACL
+	var isAuthed bool
+	var err error
+	if req.ClientToken != "" {
+		isAuthed = true
+
+		var entity *identity.Entity
+		// Load the ACL policies so we can walk the prefix for this mount
+		acl, _, entity, err = b.Core.fetchACLTokenEntryAndEntity(req)
+		if err != nil {
+			return nil, err
+		}
+		if entity != nil && entity.Disabled {
+			return nil, logical.ErrEntityDisabled
+		}
+
+	}
+
+	hasAccess := func(me *MountEntry) bool {
+		if me.Config.ListingVisibility == ListingVisibilityUnauth {
+			return true
+		}
+
+		if isAuthed {
+			return hasMountAccess(acl, me.Path)
+		}
+
+		return false
+	}
+
+	b.Core.mountsLock.RLock()
 	for _, entry := range b.Core.mounts.Entries {
-		if entry.Config.ListingVisibility == ListingVisibilityUnauth {
-			info := map[string]interface{}{
+		if hasAccess(entry) {
+			secretMounts[entry.Path] = map[string]interface{}{
 				"type":        entry.Type,
 				"description": entry.Description,
+				"options":     entry.Options,
 			}
-			secretMounts[entry.Path] = info
 		}
 	}
+	b.Core.mountsLock.RUnlock()
 
+	b.Core.authLock.RLock()
 	for _, entry := range b.Core.auth.Entries {
-		if entry.Config.ListingVisibility == ListingVisibilityUnauth {
-			info := map[string]interface{}{
+		if hasAccess(entry) {
+			authMounts[entry.Path] = map[string]interface{}{
 				"type":        entry.Type,
 				"description": entry.Description,
+				"options":     entry.Options,
 			}
-			authMounts[entry.Path] = info
 		}
 	}
+	b.Core.authLock.RUnlock()
+
+	return resp, nil
+}
+
+func (b *SystemBackend) pathInternalUIMountRead(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	path := d.Get("path").(string)
+	if path == "" {
+		return logical.ErrorResponse("path not set"), logical.ErrInvalidRequest
+	}
+	path = sanitizeMountPath(path)
+
+	me := b.Core.router.MatchingMountEntry(path)
+	if me == nil {
+		// Return a permission denied error here so this path cannot be used to
+		// brute force a list of mounts.
+		return nil, logical.ErrPermissionDenied
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"path":        me.Path,
+			"type":        me.Type,
+			"description": me.Description,
+			"options":     me.Options,
+		},
+	}
+
+	// Load the ACL policies so we can walk the prefix for this mount
+	acl, _, entity, err := b.Core.fetchACLTokenEntryAndEntity(req)
+	if err != nil {
+		return nil, err
+	}
+	if entity != nil && entity.Disabled {
+		return nil, logical.ErrPermissionDenied
```

Comment:
Should this return `logical.ErrEntityDisabled` instead?

Reply:
Yeah, should

Reply:
I was trying not to use different error message so as not to leak what mounts exist. And i wanted the ACL check to happen after making sure it's a valid mount

Reply:
Ah, ok.